Update relational-vs-nosql-data.md (#23846)
dotnet/docs: docs/architecture/cloud-native/relational-vs-nosql-data.md

Data can also be horizontally partitioned across multiple nodes, such as with [sharding](/azure/sql-database/sql-database-elastic-scale-introduction). But, sharding dramatically increases operational overhead by spitting data across many pieces that cannot easily communicate. It can be costly and time consuming to manage. Relational features that include table joins, transactions, and referential integrity require steep performance penalties in sharded deployments.

Replication consistency and recovery point objectives can be tuned by configuring whether replication occurs synchronously or asynchronously. If data replicas were to lose network connectivity in a "highly consistent" or synchronous relational database cluster, you wouldn't be able to write to the database. The system would reject the write operation as it can't replicate that change to the other data replica. Every data replica has to update before the transaction can complete.

> NoSQL databases typically support high availability and partition tolerance. They scale out horizontally, often across commodity servers. This approach provides tremendous availability, both within and across geographical regions at a reduced cost. You partition and replicate data across these machines, or nodes, providing redundancy and fault tolerance. Consistency is typically tuned through consensus protocols or quorum mechanisms - which provides finer granularity of control when navigating tradeoffs between the hard extremes typically presented with tuning synchronous vs asynchronous replication in relational systems.

Reviewer: Yes, think breaking it up makes it more readable. Let's add a hyperlink for "consensus protocols". Perhaps ==> https://blog.yugabyte.com/how-does-consensus-based-replication-work-in-distributed-databases/

> 

Author: I disagree with the original point.

Network latency isn't a unique property of NoSQL or even databases in general. The original text here seems to based on a naive assumption that all NoSQL systems are eventually consistent - which is out of touch with virtually all modern NoSQL database systems. All modern NoSQL database systems including Cosmos DB, Apache Cassandra, MongoDB have tunable quorums.

Proof:
Cosmos DB: https://docs.microsoft.com/en-us/azure/cosmos-db/consistency-levels
Apache Cassandra: https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html
MongoDB: https://docs.mongodb.com/manual/reference/read-concern/

> 

Reviewer: Good arguement and point is well taken, but, surprised by the tone, comes across as borderline-demeaning.

> 

Reviewer: I standed corrected, @aliuy brings timely information and a great deal of technical and architectural depth to the book. Kudos to @aliuy. Stayed tuned for some great updates.

> If data replicas were to lose connectivity in a "highly available" NoSQL database cluster, you could still complete a write operation to the database. The database cluster would allow the write operation and update each data replica as it becomes available. NoSQL database that support multiple writable replicas can further strengthen high availability by avoiding the need for failover when optimizing recovery time objective.

Author: I'm surprised to hear RPO and RTO are core pieces of architecture - and a key requirement for anyone whose had to implement a mission critical system. All backup and replication scenarios explain tradeoffs in terms of RPO and RTO.

> If data replicas were to lose connectivity in a "highly available" NoSQL database cluster, you could still complete a write operation to the database. The database cluster would allow the write operation and update each data replica as it becomes available. NoSQL database that support multiple writable replicas can further strengthen high availability by avoiding the need for failover when optimizing recovery time objective.

Reviewer: @aliuy - default to you. Would think it's both: The objective value and the actual recovery time itself.

> If data replicas were to lose connectivity in a "highly available" NoSQL database cluster, you could still complete a write operation to the database. The database cluster would allow the write operation and update each data replica as it becomes available. NoSQL database that support multiple writable replicas can further strengthen high availability by avoiding the need for failover when optimizing recovery time objective.

Reviewer: Let's go with this update.

> NoSQL databases typically implement partitioning (or sharding) as a native tenant to its system design. Partition management is often built-in to the database, and routing is achieved through placement hints - often called partition keys or shard keys. Flexible data models enable the NoSQL databases to lower the burden of schema management and improve availablity when deploying application updates that require data model changes.

Author: The original text here doesn't seem to understand what consistency means...

It confuses consistency in the context of ACID which has nothing to do with eventual consistency.

Eventual consistency speaks on the tradeoffs of consistency w.r.t. replication across two servers - and has nothing to do with transaction scope.

Consistency w.r.t. ACID speaks on the trade offs with respect to concurrent requests happening on the same dataset - and has nothing to do with replication (it applies to single node setups equally to replicated setups).

The document claims NoSQL systems don't support ACID transactions - which is factually incorrect. We can prove this be example using popular NoSQL database systems. How to perform ACID transactions in various NoSQL database systems:
Cosmos DB: https://devblogs.microsoft.com/cosmosdb/introducing-transactionalbatch-in-the-net-sdk/
MongoDB: https://www.mongodb.com/basics/transactions
DynamoDB: https://aws.amazon.com/blogs/aws/new-amazon-dynamodb-transactions/

I'm okay with removing the context on partitioning even though it's a core tenant of NoSQL (its the primary purpose behind the Dynamo, BigTable, Cassandra, HBase, et al projects) - but the original text should be removed because it is factually incorrect. Also think its a big miss to write an article on NoSQL and miss the primary reason for why many NoSQL systems were created in the first place (including why implemented Cosmos DB within Microsoft).

> NoSQL databases typically implement partitioning (or sharding) as a native tenant to its system design. Partition management is often built-in to the database, and routing is achieved through placement hints - often called partition keys or shard keys. Flexible data models enable the NoSQL databases to lower the burden of schema management and improve availablity when deploying application updates that require data model changes.

Reviewer: Appreciate @aliuy's depth in NoSQL technologies. Think the best thing is for @aliuy and I to get on a call and craft the excat change. I'll reach out to him.

> NoSQL databases typically implement partitioning (or sharding) as a native tenant to its system design. Partition management is often built-in to the database, and routing is achieved through placement hints - often called partition keys or shard keys. Flexible data models enable the NoSQL databases to lower the burden of schema management and improve availablity when deploying application updates that require data model changes.

Reviewer: Talked with @aliuy.  Let's go with the following:   

Moden NoSQL databases typically implement partitioning capabilities as a feature of their system design. Partition management is often built-in to the database, and routing is achieved through placement hints - often called partition keys. A flexible data models enables the NoSQL databases to lower the burden of schema management and improve availablity when deploying application updates that require data model changes.****

> NoSQL databases typically implement partitioning (or sharding) as a native tenant to its system design. Partition management is often built-in to the database, and routing is achieved through placement hints - often called partition keys or shard keys. Flexible data models enable the NoSQL databases to lower the burden of schema management and improve availablity when deploying application updates that require data model changes.

Reviewer: ```suggestion
Moden NoSQL databases typically implement partitioning capabilities as a feature of their system design. Partition management is often built-in to the database, and routing is achieved through placement hints - often called partition keys. A flexible data models enables the NoSQL databases to lower the burden of schema management and improve availablity when deploying application updates that require data model changes.
```

> High availability and massive scalability are often more critical to the business than relational table joins and referential integrity. Developers can implement techniques and patterns such as Sagas, CQRS, and asynchronous messaging to embrace eventual consistency.

Author: Strong consistency doesn't make sense in this context per above.

> High availability and massive scalability are often more critical to the business than relational table joins and referential integrity. Developers can implement techniques and patterns such as Sagas, CQRS, and asynchronous messaging to embrace eventual consistency.

Reviewer: Great point. Let's accept change.

|  Consider a NoSQL datastore when: | Consider a relational database when: |
| :-------- | :-------- |

> | You have high volume workloads that require predictable latency at large scale (e.g. latency measured in milliseconds while performing millions of transactions per second) | Your workload volume generally fits within thousands of transactions per second |

Reviewer: Please add line two back in: The ACID guarantees. Let's add change with the ACID guarantees line added back.

Author: ACID is not a differentiator between NoSQL vs RDBMS per above. I'm guessing this was a miss in which the original author hasn't had experience with modern NoSQL databases created in the last 10 years.

Reviewer: See the reader's point of view and agree on the change. But, again, disappointed by the sarcastic, demeaning tone.

Reviewer: I stand corrected. Totally misunderstood the context. @aliuy explained the concept of gen1 and gen2 NoSQL datastores. Excited that @aliuy is bringing timely information and a great deal of technical and architectural depth to the book. Kudos to @aliuy. Stayed tuned for some great updates.

![Overview of Cosmos DB](./media/cosmos-db-overview.png)

> **Figure 5-12**: Overview of Azure Cosmos DB

Reviewer: Yes, please accept this change.

|  | Azure Table Storage  | Azure Cosmos DB  |
| :-------- | :-------- |:-------- |
| Latency | Fast | Single-digit millisecond latency for reads and writes anywhere in the world |

> | Throughput | Limit of 20,000 operations per table | Unlimited operations per table |

Reviewer: Yes, please accept change.

| Global Distribution | Single region with optional single secondary read region | Turnkey distributions to all regions with automatic failover |
| Indexing | Available for partition and row key properties only | Automatic indexing of all properties |

> | Pricing | Optimized for cold workloads (low throughput : storage ratio) | Optimizied for hot workloads (high throughput : storage ratio) |

Reviewer: Yes, please accept change.

- Azure Database for MySQL
- Azure Database for MariaDB
- Azure Database for PostgreSQL

> - Azure Cosmos DB

Reviewer: Please accept change.
